### PR TITLE
add basic SQS consumer functionality

### PR DESCRIPTION
**Pull request by adriangb**
Initial draft for implementation and testing plan. @samuelcolvin I'd appreciate an initial review and then I can dig deeper into testing edge cases, getting the queue URL, etc.

### CODE REVIEW

File: aioaws/core.py [hunk 1]
```diff
@@ -86,140 +189,84 @@ async def request(
...
         error_msg = f'unexpected response from {r.request.method} "{r.request.url}": {r.status_code}'
         super().__init__(error_msg)
         self.response = r
         self.status = r.status_code
 
     def __str__(self) -> str:
         return f'{self.args[0]}, response:\n{self.response.text}'
+
+
+class AWSV4AuthFlow(httpx.Auth):
```
Comment (adriangb) [inline aioaws/core.py:249]: The `AWSClient` class was going to be difficult to adapt to SQS, e.g. because the host name for some requests depends on the queue URL while others depend on the service/region. So this is an attempt to decouple the auth portion of things from the hostname / other stuff going on in AWSClient. I can try to find a way to work around things instead if you'd prefer.

File: aioaws/sqs.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            },
+            auth=self._auth,
+            headers={
+                "Accept": "application/json",
+            },
+        )
+
+
+@asynccontextmanager
+async def create_sqs_client(
```
Comment (adriangb) [inline aioaws/sqs.py:163]: The idea here is to avoid the state of never running `AsyncClient.__aexit__`. This can be done using `SQSClient.__aexit__` but it's not transmittable via the type system and more prone to user error (you can instantiate `SQSClient` without using it as a context manager).

File: aioaws/core.py [hunk 3]
```diff
@@ -23,12 +23,108 @@
...
+        aws_access_key_id: str,
+        region: str,
+        service: str,
+    ) -> None:
+        self.aws_secret_key = aws_secret_key
+        self.aws_access_key_id = aws_access_key_id
+        self.region = region
+        self.service = service
+
+    def auth_headers(
```
Comment (samuelcolvin) [inline aioaws/core.py:51]: This logic is very specific and very small changes can break authentication in opaque ways.

Any chance you could move `_AWSv4Auth` below `AwsClient` so the diff is smaller and all this code remains (in git's eyes) unchanged? No worries if not.

File: aioaws/core.py [hunk 4]
```diff
@@ -5,10 +5,10 @@
 from binascii import hexlify
 from datetime import datetime
 from functools import reduce
-from typing import TYPE_CHECKING, Any, Dict, List, Literal, Optional, Tuple
+from typing import TYPE_CHECKING, Any, Dict, Generator, List, Literal, Optional, Tuple
 from urllib.parse import quote as url_quote
 
-from httpx import URL, AsyncClient, Response
+import httpx
```
Comment (samuelcolvin) [inline aioaws/core.py:11]: Sorry to be be boring, let's reverse this to minimise the changes.

Comment (samuelcolvin) [reply aioaws/core.py:11]: Looks like `Request` is the only other thing you need?

File: aioaws/core.py [hunk 5]
```diff
@@ -23,12 +23,108 @@
...
+
+def _aws4_x_amz_date(dt: datetime) -> str:
+    return dt.strftime('%Y%m%dT%H%M%SZ')
+
+
+def _aws4_reduce_signature(key: bytes, msg: str) -> bytes:
+    return hmac.new(key, msg.encode(), hashlib.sha256).digest()
+
+
+class _AWSv4Auth:
```
Comment (samuelcolvin) [inline aioaws/core.py:38]:
```suggestion
class AWSv4Auth:
```

We don't need to add it to `__all__`, but I think we can remove the `_` shouting "THIS IS VERY PRIVATE".

File: aioaws/sqs.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                    self._client,
+                    auth=self._auth,
+                )
+            )
+        return self._queue_name_or_url.url
+
+    async def poll(
+        self,
+        *,
+        config: PollConfig = PollConfig(),
```
Comment (samuelcolvin) [inline aioaws/sqs.py:96]: mutable default, better to do `Optional[..] = None ... config = PollConfig()`?

**Review by samuelcolvin — COMMENTED** [review]
Mostly I think this looks great, a part from the specifics:
* let's try to keep the diff to a minimum, partly to help reviews now and partly to make tracing changes easier in future
* please add docstrings and comments as much as possible (I know I'm bad at this, but still better to do so now)
* you'll need to add docs to the readme

**Comment by samuelcolvin** [comment]
Overall, this looks great, thank you.

File: aioaws/sqs.py (hunk 6 above)
Comment (adriangb) [reply aioaws/sqs.py]: addressed

File: aioaws/core.py (hunk 4 above)
Comment (adriangb) [reply aioaws/core.py]: addressed

File: aioaws/core.py (hunk 5 above)
Comment (adriangb) [reply aioaws/core.py]: addressed

File: aioaws/core.py (hunk 3 above)
Comment (adriangb) [reply aioaws/core.py]: Agreed. I didn't make any changes FWIW. I moved things around to try to minimize the diff.

**Comment by adriangb** [comment]
Feedback addressed. Still have to:
- Write some more tests for edge cases and `ack`/`delete` features
- Do a smoke test against a real SQS queue to make sure I'm making the API calls correctly

**Comment by adriangb** [comment]
all done. smoke tests helped me fix the API calls. I'm not including them because they're super crude, but given the stability of the HTTP API I think we should be fine just ensuring we don't break what's there right now. What do you think @samuelcolvin, is this an acceptable testing strategy?

**Comment by samuelcolvin** [comment]
Yes I think so the aws4 Auth is tested against aws on commits to master.

I'll look in detail on Monday.

**Comment by samuelcolvin** [comment]
Thanks so much for this. Sorry it took so long.

**Comment by adriangb** [comment]
No worries, I'm very glad this is in now!

**Comment by samuelcolvin** [comment]
I'll do a new release soon. Just checking this works properly with CloudFlare's R2.
